apps/notifications/jobs.py

```diff
@@ -0,0 +1,88 @@
+from sqlalchemy import and_
+from main import db
+from datetime import datetime, timedelta
+from flask import current_app as app
+
+from models import scheduled_task
+from models.cfp import FavouriteProposal, Proposal
+from models.user import User
+from models.web_push import PushNotificationJob
+from models.notifications import UserNotificationPreference
+from pywebpush import webpush, WebPushException
+
+
+def deliver_notification(job: PushNotificationJob):
+    """Deliver a push notification from a PushNotificationJob.
+
+    The passed job will be mutated to reflect delivery state. A job which isn't
+    queued will be skipped over.
+    """
+    if job.state != "queued":
+        return
+
+    try:
+        webpush(
+            subscription_info=job.target.subscription_info,
+            data=job.title,
+            vapid_private_key=app.config["WEBPUSH_PRIVATE_KEY"],
+            vapid_claims={
+                "sub": "mailto:contact@emfcamp.org",
+            },
+        )
+
+        job.state = "delivered"
+    except WebPushException as err:
+        job.state = "failed"
+        job.error = err.message
+
+
+@scheduled_task(minutes=1)
+def send_queued_notifications():
+    jobs = PushNotificationJob.query.where(
+        PushNotificationJob.state == "queued"
+        and (PushNotificationJob.not_before is None or PushNotificationJob.not_before <= datetime.now())
+    ).all()
+
+    for job in jobs:
+        deliver_notification(job)
+        db.session.add(job)
+
+    db.session.commit()
+
+
+@scheduled_task(minutes=15)
+def queue_content_notifications(time=None) -> None:
```

Comment:
I deeply dislike the approach being taken here, if anyone with sufficient SQLAlchemy knowledge wants to turn it into a single query with some joins I'd be very happy.